README.md
mpimessagespace
===============

This tool shows the "send" messages of a set of MPI processes in a 3D spaces. It draws each "send" as a point. The x-axis is the time, the y-axis is the message's size, and the processes are on z-axis. All points related to a single process are in a plane parallel to the x-y one.

Keyboard commands
-----------------

    *Q*: quit;
    *Arrows*: rotate the axis;
    *A/D/W/S/Pag up/Pag down*: move the camera;
    *P*: show/hide process planes;
    *+/-*: when showing process planes, choose the highlighted one;
    *T*: fill the graph with random test data.

![Screenshot 1](screenshots/scr1.png)
![Screenshot 2](screenshots/scr2.png)
![Screenshot 3](screenshots/scr3.png)
![Screenshot 4](screenshots/scr4.png)
![Screenshot 5](screenshots/scr5.png)


src/NEWS
2009 - Development
2013 - GitHub release


src/main.c
/*
 * $Id: main.c 901 2009-05-07 08:50:35Z epmancini $
 */
#ifdef WIN32
#include <windows.h>
#include <GL/glut.h>
#endif

#ifdef __APPLE__
#include <GLUT/glut.h>
#endif

#ifdef __linux__
#include <GL/glut.h>
#endif

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "messagespace.h"
#include "datafile.h"


///
/// Test data
///
void testData() {
    int p,i;
    countPoints = 0;
    numProcs    = 4;
    numpoints   = (int *)malloc( sizeof(int) * numProcs );
    points      = (pPoint *)  malloc( sizeof(pPoint)   * numProcs );
    for (p=0; p<numProcs; p++){
        numpoints[p] = 1000+p*100;
        points[p]=malloc( sizeof(Point) * numpoints[p] );
        for (i=0; i<numpoints[p]; i++){
            points[p][i][0] = (GLdouble)i;
            points[p][i][1] = ((GLdouble)rand())*(double)i/(double)RAND_MAX;
            points[p][i][2] = (GLdouble)p;
            countPoints++;
        }
    }
}


///
/// Handle keyboard's special keys
///
void keyboardspecial(int key, int x, int y)
{
    const GLfloat ANGLE_DELTA = 10.0;
    switch (key)  {
        case GLUT_KEY_DOWN:
            angleX -= ANGLE_DELTA;
            if (angleX<0) angleX=360;
            draw();
            break;
        case GLUT_KEY_UP:
            angleX += ANGLE_DELTA;
            if (angleX>360) angleX=0;
            draw();
            break;
        case GLUT_KEY_LEFT:
            angleY -= ANGLE_DELTA;
            if (angleY<0) angleY=360;
            draw();
            break;
        case GLUT_KEY_RIGHT:
            angleY += ANGLE_DELTA;
            if (angleY>360) angleY=0;
            draw();
            break;
        case GLUT_KEY_PAGE_UP:
            cameraY += 0.01f;
            draw();
            break;
        case GLUT_KEY_PAGE_DOWN:
            cameraY -= 0.01f;
            draw();
            break;
    }
}

///
/// Handle keyboard's keys
///
void keyboard(unsigned char key, int x, int y)
{
  switch (key)
  {
    case 't':
    case 'T':
          testData();
          normalizeData();
            glutPostRedisplay();
          break;

    case '+':
            selectedProcessPlanes++;
            if (selectedProcessPlanes>=numProcs)
                selectedProcessPlanes = -1;
            glutPostRedisplay();
            break;

    case '-':
            selectedProcessPlanes--;
            if (selectedProcessPlanes<-1) 
                selectedProcessPlanes = numProcs-1;
            glutPostRedisplay();
            break;

    case 'w':
    case 'W':
            cameraZ += 0.01f;
            glutPostRedisplay();
            break;

    case 'z':
    case 'Z':
            if (zyPlane_x>0.0f) {
                zyPlane_x -= 0.01f;
                glutPostRedisplay();
            }
            break;

    case 'x':
    case 'X':
            if (zyPlane_x<1.0f) {
                zyPlane_x += 0.01f;
                glutPostRedisplay();
            }
            break;

    case 's':
    case 'S':
            cameraZ -= 0.01f;
            glutPostRedisplay();
            break;

    case 'a':
    case 'A':
            cameraX -= 0.01f;
            glutPostRedisplay();
            break;

    case 'd':
    case 'D':
            cameraX += 0.01f;
            glutPostRedisplay();
            break;

    case 'p':
    case 'P':
        procsPlanes = !procsPlanes;
        glutPostRedisplay();
        break;

    case 0x1B:
    case 'q':
    case 'Q':
        exit(0);
        break;
  }
}

///
/// Handle mouse events
///
void mouse(int btn, int state, int x, int y)
{
    /*
  if (state == GLUT_DOWN)
  {
    if (btn == GLUT_LEFT_BUTTON)
      angle = angle + 1.0f;
    else if (btn == GLUT_RIGHT_BUTTON)
      angle = angle - 1.0f;
    else
      angle = 0.0f;
  }
  */
}


///
/// Main
///
int  main( int argc, char **argv ){
    int i;
    glutInit(&argc, argv);
    init();

    printf("Loading data.\n");
    numProcs = argc-1;
    numpoints   = (int *)malloc( sizeof(int) * numProcs );
    points      = (pPoint *)  malloc( sizeof(pPoint)   * numProcs );
    for (i=1; i<argc; i++) {
        printf(" %d) %s: ", i, argv[i]);
        loadFile(i-1, argv[i]);
        printf("done\n");
    }

    //testData();
    printf("  %u messages.\n", countPoints);

    printf("Normalizing data...\n");
    normalizeData();

    printf("  Time: [%f, %f]\n", minT, maxT );
    printf("  Size: [%f, %f]\n", minDim, maxDim );
    printf("  Processes: [%d, %d]\n", (int)minProc, (int)maxProc );

    printf("Done.\n");

    // Window init
    glutInitWindowSize(800, 600);
    glutInitDisplayMode(GLUT_RGB | GLUT_DOUBLE);
    glutCreateWindow("Messagespace");
    glutKeyboardFunc(keyboard);
    glutSpecialFunc( keyboardspecial );
    glutDisplayFunc(draw);
    //glutMouseFunc(mouse);
    //glutIdleFunc(drawcube); 

    // OpenGl init
    glMatrixMode(GL_PROJECTION);
    glLoadIdentity();
    glOrtho( gl_x0, gl_x1, gl_y0, gl_y1, gl_z0, gl_z1);
    glHint(GL_LINE_SMOOTH_HINT, GL_NICEST );
    glHint(GL_POINT_SMOOTH_HINT, GL_NICEST );
    glHint(GL_PERSPECTIVE_CORRECTION_HINT, GL_NICEST );
    glBlendFunc(GL_SRC_ALPHA, GL_ONE_MINUS_SRC_ALPHA);

    glMatrixMode(GL_MODELVIEW);
    glClearColor(0.0, 0.0, 0.0, 0.0);

    glutMainLoop();
    return 0;
}


src/datafile.c
/*
 * $Id: datafile.c 891 2009-05-04 10:38:49Z epmancini $
 */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "messagespace.h"
#include "datafile.h"

#define BUFSIZE 256

/**
 * Not yet implemented
 */
int isVoidOrComment(char *buf) {
    return 0;
}

/**
 * Load a datafile. Format: 
 *   from to size time
 */
void loadFile(int p, char *name) {
    FILE *f;
    char buf[BUFSIZE];
    int count = 0;

    f = fopen(name, "rt");
    // Count lines
    while (!feof(f)) {
        fgets(buf, BUFSIZE, f);
        if (!isVoidOrComment(buf)) {
            count++;
        }
    }

    countPoints+=count;
    numpoints[p] = count;
    points[p] = malloc( sizeof(Point) * count );

    fseek( f, 0L, SEEK_SET );
    count = 0;
    while (!feof(f)) {
        int from, to, size;
        double time;
        from = to = size = 0;
        time = 0;
        fgets(buf, BUFSIZE, f);
        
        

        if (!isVoidOrComment(buf)) {
            sscanf(buf, "%d %d %d %lf\n", &from, &to, &size, &time );
            /*
            time -= 1168956416.0;
            printf( " %s", buf );
            printf( " %d %d %d %lf\n", from, to, size, time );
            */
            points[p][count][0] = (double)(time);
            points[p][count][1] = (double)size;
            points[p][count][2] = (double)p;
            count++;
        }
    }

    fclose(f);
}


src/messagespace.h
/*
 * $Id: messagespace.h 901 2009-05-07 08:50:35Z epmancini $
 */
#ifndef MESSAGESPACE_H
#define MESSAGESPACE_H

/// Points
typedef double Point[3];
typedef Point *pPoint;

extern Point **points;
extern int *numpoints;
extern int numProcs;
extern int showProcessPlanes;
extern int selectedProcessPlanes;

/// Total number of points
extern unsigned countPoints;

/// Rotation angles
extern float angleX;
extern float angleY;
extern float angleZ;

/// Move camera
extern float cameraX, cameraY, cameraZ;

/// Opengl orthogonal box that includes graph
extern float gl_x0, gl_x1, gl_y0, gl_y1, gl_z0, gl_z1;

/// Length of axis
extern float axis_x_len, axis_y_len, axis_z_len;

/// Minimum and maximum time: minimum is reported to 0 on axis, maximum to 1
extern double  minT, maxT;

extern float zyPlane_x;

/// First an last processes: min->0 max->1
extern double  minProc, maxProc;

/// Minimum and maximum of message size: min->0 max->1
extern double minDim, maxDim;

extern int procsPlanes;


void init();
void normalizeData();
void draw(void);

#endif


src/messagespace.c
/*
 * $Id: messagespace.c 901 2009-05-07 08:50:35Z epmancini $
 */
#ifdef WIN32
#include <windows.h>
#include <GL/glut.h>
#endif

#ifdef __linux__
#include <GL/glut.h>
#endif

#ifdef __APPLE__
#include <GLUT/glut.h>
#endif

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "messagespace.h"
#include "datafile.h"

float strip_width = 0.01f;

/// Colors
float COLOR_TEXT[]              = {1.0f, 1.0f, 1.0f};
float COLOR_POINTS[]            = {1.0f, 1.0f, 0.2f};
float COLOR_UNSELECTED_POINTS[] = {0.5f, 0.4f, 0.0f};
float COLOR_AXIS[]              = {0.7f, 0.7f, 1.0f};
float COLOR_AXIS_LABELS[]       = {0.8f, 0.8f, 1.0f};
float COLOR_PROCESS_STRIPS[]    = {0.4f, 0.4f, 0.5f};

float COLOR_PROCESS_PLANE[]     = {0.2f, 0.8f, 0.2f, 0.2f};
float COLOR_ZY_PLANE[]          = {0.2f, 0.2f, 0.8f, 0.5f};

#define HELP_STRING "Q: Exit - T: test data - P: process planes - +/-: select plane - A/D/W/S/Pag up/Pag down: move camera - Arrows: rotate "

/// Minimum and maximum time: minimum is reported to 0 on axis, maximum to 1
double minT, maxT;

/// First an last processes: min->0 max->1
double minProc, maxProc;

/// Minimum and maximum of message size: min->0 max->1
double minDim, maxDim;

/// Rotation angles
float angleX=0.0f;
float angleY=0.0f;
float angleZ=0.0f;

float cameraX=0.0f, cameraY=0.0f, cameraZ=0.0f;

/// Opengl orthogonal box that includes graph
float gl_x0, gl_x1, gl_y0, gl_y1, gl_z0, gl_z1;

/// Length of axis
float axis_x_len, axis_y_len, axis_z_len;

Point **points      = NULL;
int *numpoints      = NULL;
int numProcs        = 0;
int procsPlanes     = 0;

/// Total number of points
unsigned countPoints = 0;

int selectedProcessPlanes = -1;

void draw_axes(void);
void output(GLfloat x, GLfloat y, char *s);

float zyPlane_x;

///
/// Initialization
///
void init() {
    zyPlane_x = 0.0f;
    angleX  = 340.0f;
    angleY  = 340.0f;
    angleZ  = 0.0f;
    cameraX = 0.2f;
    cameraY = cameraZ = 0.0f;
    
    axis_x_len = 1.0f;
    axis_y_len = 1.0f;
    axis_z_len = 1.0f;
    
    gl_x0 = gl_y0 = -0.5f;
    gl_x1 = gl_y1 = 1.5;
    
    gl_z0 = -5;
    gl_z1 = 5;

    selectedProcessPlanes = -1;

}

///
/// Data normalization
///
void normalizeData() {
    int p, i;

    if (numProcs<=0)
        return;
    
    minT    = maxT    = points[0][0][0];
    minDim  = maxDim  = points[0][0][1];
    minProc = maxProc = points[0][0][2];
    for (p=0; p<numProcs; p++){
        for (i=0; i<numpoints[p]; i++){
            if (points[p][i][0]<minT) {
                minT = points[p][i][0];
            } else if (points[p][i][0]>maxT) {
                maxT = points[p][i][0];
            }
            if (points[p][i][1]<minDim) {
                minDim = points[p][i][1];
            } else if (points[p][i][1]>maxDim) {
                maxDim = points[p][i][1];
            }
            if (points[p][i][2]<minProc) {
                minProc = points[p][i][2];
            } else if (points[p][i][2]>maxProc) {
                maxProc = points[p][i][2];
            }
        }
    }

    for (p=0; p<numProcs; p++){
        for (i=0; i<numpoints[p]; i++){
            points[p][i][0] = (points[p][i][0]-minT)    / (maxT-minT);
            points[p][i][1] = (points[p][i][1]-minDim)  / (maxDim-minDim);
            points[p][i][2] = (points[p][i][2]-minProc) / (maxProc-minProc);
            points[p][i][2] = points[p][i][2] * (1.0f-2.0f*strip_width)+strip_width;
        }
    }
}

///
/// Draw a string
///
void output(GLfloat x, GLfloat y, char *s)
{
  //int len, i;
  glRasterPos2f(x, y);
  //len = (int) strlen(string);
  //for (i = 0; i < len; i++)
  while (*s != 0)
  {
      glutBitmapCharacter(GLUT_BITMAP_HELVETICA_10, *s);
      s++;
  }
}

void output3(GLfloat x, GLfloat y, GLfloat z, char *s)
{
  //int len, i;
  glRasterPos3f(x, y, z);
  //len = (int) strlen(string);
  //for (i = 0; i < len; i++)
  while (*s != 0)
  {
      glutBitmapCharacter(GLUT_BITMAP_HELVETICA_10, *s);
      s++;
  }
}

///
/// Draw cartesian axes
///
void draw_axes(void) {
  const GLfloat ARROW_W = 0.02f;
  int p;

  glColor3fv(COLOR_AXIS);
  glBegin(GL_LINES);
    glVertex3f(0.0f, 0.0f, 0.0f);
    glVertex3f(axis_x_len, 0.0f, 0.0f);
    glVertex3f(axis_x_len, 0.0f, 0.0f);
    glVertex3f(axis_x_len-ARROW_W, ARROW_W, 0.0f);
    glVertex3f(axis_x_len, 0.0f, 0.0f);
    glVertex3f(axis_x_len-ARROW_W, -ARROW_W, 0.0f);
    
    glVertex3f(0.0f, 0.0f, 0.0f);
    glVertex3f(0.0f, axis_y_len, 0.0f);
    glVertex3f(0.0f, axis_y_len, 0.0f);
    glVertex3f(ARROW_W, axis_y_len-ARROW_W, 0.0f);
    glVertex3f(0.0f, axis_y_len, 0.0f);
    glVertex3f(-ARROW_W, axis_y_len-ARROW_W, 0.0f);
    
    glVertex3f(0.0f, 0.0f, 0.0f);
    glVertex3f(0.0f, 0.0f, axis_z_len);
    glVertex3f(0.0f, 0.0f, axis_z_len);
    glVertex3f(ARROW_W, 0.0f, axis_z_len-ARROW_W);
    glVertex3f(0.0f, 0.0f, axis_z_len);
    glVertex3f(-ARROW_W, 0.0f, axis_z_len-ARROW_W);
  glEnd();

  glBegin(GL_QUADS);
  glColor3fv(COLOR_PROCESS_STRIPS);
  for (p=0; p<numProcs; p++) {
      float z = (float)points[p][0][2];
      float stripWidth = 0.01f;
      glVertex3f(0.0f, 0.0f, z-stripWidth );
      glVertex3f(axis_x_len, 0.0f, z-stripWidth);
      glVertex3f(axis_x_len, 0.0f, z+stripWidth);
      glVertex3f(0.0f, 0.0f, z+stripWidth);
  }
  glEnd();
  
  for (p=0; p<numProcs; p++) {
      char s[32];
      _itoa(p, s, 10 );
      output3(-2*ARROW_W, 0.0f, (float)points[p][0][2], s );
  }

  glColor3fv(COLOR_AXIS_LABELS);
  output(axis_x_len, 0.0f-ARROW_W, "t");
  output(0.0f-ARROW_W, axis_y_len, "size");
}


///
/// Draw the send points
///
void draw(void)
{
    int  p, i;
    GLfloat BORDER = 0.02f;

    glClear(GL_COLOR_BUFFER_BIT|GL_DEPTH_BUFFER_BIT);

    glMatrixMode(GL_MODELVIEW);
    glLoadIdentity();
    glEnable(GL_DEPTH_TEST);
    glColor3fv( COLOR_TEXT );
    output(gl_x0+BORDER, gl_y0+BORDER, HELP_STRING);

    glRotatef(angleX, 1.0f, 0.0f, 0.0f); 
    glRotatef(angleY, 0.0f, 1.0f, 0.0f); 
    glRotatef(angleZ, 0.0f, 0.0f, 1.0f); 
    glTranslatef(cameraX, cameraY, cameraZ );
    draw_axes();



    glBegin(GL_POINTS);
    glColor3fv(COLOR_POINTS);
    for (p=0; p<numProcs; p++) {
        // Colors for a process
        if ((selectedProcessPlanes>=0) && (procsPlanes)) {
            if (p!=selectedProcessPlanes) {
                glColor3fv(COLOR_UNSELECTED_POINTS);
            } else {
                glColor3fv(COLOR_POINTS);
            }
        }
        // Ponint for a single process
        for (i=0; i<numpoints[p]; i++) {
            glVertex3dv(points[p][i]);
        }
    }
    glEnd();

    if (procsPlanes) {
      //glDisable(GL_DEPTH_TEST);
      glEnable(GL_BLEND);
      glBegin(GL_QUADS);
      glColor4fv(COLOR_PROCESS_PLANE);
      for (p=numProcs-1; p>=0; p--) {
          if ( (selectedProcessPlanes<0) || (selectedProcessPlanes==p) ) {
              glVertex3f(0.0f, 0.0f, (GLfloat)points[p][0][2]);
              glVertex3f(0.0f, axis_y_len, (GLfloat)points[p][0][2]);
              glVertex3f(axis_x_len, axis_y_len, (GLfloat)points[p][0][2]);
              glVertex3f(axis_x_len, 0.0f, (GLfloat)points[p][0][2]);
          }
      }
      glEnd();
      glDisable(GL_BLEND);
    }

    /*
    if ((zyPlane_x>0.0f) && (zyPlane_x<1.0f)) {
      //glDisable(GL_DEPTH_TEST);
      glEnable(GL_BLEND);
      glBegin(GL_QUADS);
        glColor4fv(COLOR_ZY_PLANE);
        glVertex3f(zyPlane_x, 0.0f,       0.0f );
        glVertex3f(zyPlane_x, axis_y_len, 0.0f);
        glVertex3f(zyPlane_x, axis_y_len, axis_z_len);
        glVertex3f(zyPlane_x, 0.0f,       axis_z_len);
      glEnd();
      glDisable(GL_BLEND);
    }
    */

    glFlush();
    glutSwapBuffers();
}

